### ISSUE TITLE
Is this still useful since Rust 1.78 enabled `unsafe` stdlib precondition assertions?

### ISSUE THREAD

**Issue by langston-barrett** [RalfJung/cargo-careful#30]
Hi, first of all thanks for all your work on this tool and on Rust in general!

I read in [the Rust 1.78 release blog post](https://blog.rust-lang.org/2024/05/02/Rust-1.78.0.html#asserting-unsafe-preconditions) that the stdlib now enables `unsafe` precondition checks depending on the *user's* setting of `debug_assertions`:

> The Rust standard library has a number of assertions for the preconditions of unsafe functions, but historically they have only been enabled in #[cfg(debug_assertions)] builds of the standard library to avoid affecting release performance. However, since the standard library is usually compiled and distributed in release mode, most Rust developers weren't ever executing these checks at all.
> 
> Now, the condition for these assertions is delayed until code generation, so they will be checked depending on the user's own setting for debug assertions -- enabled by default in debug and test builds.

The README for `cargo-careful` says:

> The most important thing cargo careful does is that it builds the standard library with debug assertions.

Is this tool still useful in Rust >1.78? In either case, it might be helpful to call this out in the README and explain why or why not.

**Comment by RalfJung** [comment]
Yeah that's a good question. This does still enable some more checks, like `-Zextra-const-ub-checks -Zstrict-init-checks`, but it may be basically equivalent to setting these things in RUSTFLAGS.

@saethlin you probably know better -- is there still any reason to build the standard library with debug assertions?

**Comment by saethlin** [comment]
> Now, the condition for these assertions is delayed until code generation, so they will be checked depending on the user's own setting for debug assertions -- enabled by default in debug and test builds.

Ah, sorry I didn't proofread this well enough. This statement in the blog is technically right but morally wrong. Some of the (delayed) checks are still cfg'd out because their impact on compile times exceeds the thresholds that the perf wg established for significant regressions. For example: https://github.com/rust-lang/rust/blob/c7716d543191e52ed817b725f2fab54bbd5d2707/library/core/src/ptr/mod.rs#L1316-L1317

And there are still plenty of plain old debug assertions in the standard library which do things like validate internal consistency of data structures. All such checks implemented with `cfg(debug_assertions)` or `debug_assert!` are still stripped out of the distributed standard library.

**Comment by RalfJung** [comment]
> And there are still plenty of plain old debug assertions in the standard library which do things like validate internal consistency of data structures.

Sure, but those would indicate a bug in the standard library, not the user's code.

Those UB checks that are still compiled out seem to be the main difference then. We should clarify that in the readme.
